fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/expressions/Comparisons.java

```diff
@@ -964,18 +1043,26 @@ public String getParameter() {
             return parameter;
         }
 
+        @Nonnull
+        public CorrelationIdentifier getAlias() {
+            if (internal != Bindings.Internal.CORRELATION) {
+                throw new IllegalStateException("caller should check for type of binding before calling this method");
+            }
+            return CorrelationIdentifier.of(Bindings.Internal.CORRELATION.identifier(parameter));
+        }
+
         @Override
+        @SpotBugsSuppressWarnings("EQ_UNUSUAL")
+        @SuppressWarnings("EqualsWhichDoesntCheckParameterClass")
         public boolean equals(Object o) {
-            if (this == o) {
-                return true;
-            }
-            if (o == null || getClass() != o.getClass()) {
-                return false;
+            final AliasMap aliasMap;
+            if (internal == Bindings.Internal.CORRELATION) {
+                final var alias = CorrelationIdentifier.of(Bindings.Internal.CORRELATION.identifier(parameter));
+                aliasMap = AliasMap.identitiesFor(ImmutableSet.of(alias));
+            } else {
+                aliasMap = AliasMap.emptyMap();
             }
```

Comment:
Do we not expect `equals` to be done very often on / around these? Because this is now doing allocation.

Reply:
This is already done in a lot of `structuralEquals()` and `semanticEquals()` methods. I can make `emptyMap()` return a static instance so the general (the old-planner code path) is not affected by this at all as that one should never create an actual correlation. I can also minimize the allocations by creating a special subclass of `AliasMap()` that just implements identities thus there won't be any  need for a copy of the aliases.

Reply:
Made `emptyMap()` return a neutral element. In this way the old-planner code path is not punished.